comp/core/workloadmeta/collectors/internal/kubelet/kubelet.go

```diff
@@ -444,24 +430,14 @@ func extractResources(spec *kubelet.ContainerSpec) workloadmeta.ContainerResourc
 	}
 
 	// extract GPU resource info from the possible GPU sources
-	uniqueGPUVendor := make(map[string]bool)
-
-	resourceKeys := make([]kubelet.ResourceName, 0, len(spec.Resources.Requests))
+	uniqueGPUVendor := make(map[string]struct{})
 	for resourceName := range spec.Resources.Requests {
-		resourceKeys = append(resourceKeys, resourceName)
-	}
-
-	for _, gpuResourceName := range kubelet.GetGPUResourceNames() {
-		for _, resourceKey := range resourceKeys {
-			if strings.HasPrefix(string(resourceKey), string(gpuResourceName)) {
-				if gpuReq, found := spec.Resources.Requests[resourceKey]; found {
-					resources.GPURequest = pointer.Ptr(uint64(gpuReq.Value()))
-					uniqueGPUVendor[extractGPUVendor(gpuResourceName)] = true
-					break
-				}
-			}
+		gpuName, found := gpu.ExtractSimpleGPUName(gpu.ResourceGPU(resourceName))
+		if found {
+			uniqueGPUVendor[gpuName] = struct{}{}
 		}
 	}
+
```

Comment:
can we wrap this logic in one function to be reused in `comp/core/workloadmeta/collectors/util/kubernetes_resource_parsers/pod.go` ?
I feel like the code is very similar

Reply:
I've tried implementing a function such as the following:
```
func GetGPUResourcesFromResourceList[T corev1.ResourceList | kubelet.ResourceList](gpuSet *map[string]struct{}, resourceList T) {
	for resourceName := range resourceList {
		gpuName, found := ExtractSimpleGPUName(ResourceGPU(resourceName))
		if found {
			(*gpuSet)[gpuName] = struct{}{}
		}
	}
}
```
However, I get an error stating "cannot range over resourceList: no core type". It seems that Go's type system cannot infer the underlying map structure directly from `T corev1.ResourceList | kubelet.ResourceList`.
I've tried directly setting it to `[T ~map[corev1.ResourceName]resource.Quantity | ~map[kubelet.ResourceName]resource.Quantity]` and get the same issue. It's interesting that `corev1.ResourceName` and `kubelet.ResourceName` are both simply strings, so these types are essentially the same. I tried type casting the input `map[string]resource.Quantity(container.Resources.Requests)` but it doesn't allow me either.

I can't think of a lightweight way for wrapping this logic into one function that can be reused. Do you have any ideas? It would be very appreciated. Thanks!